### PR TITLE
feat: support import maps in functions serve and deploy

### PR DESCRIPTION
**Pull request by laktek**
## What kind of change does this PR introduce?

This adds support for import maps in `function serve` and `function deploy`. 

### What are Import Maps?

https://deno.land/manual@v1.28.3/basics/modules/import_maps

For deployed functions, `import_map.json` should always be placed in function's top-level directory.

### CODE REVIEW

**Comment by soedirgo** [comment]
Can we make the `import_map.json` path non-configurable? It's a bit odd that it can be set for `serve` but not `deploy`

**Comment by laktek** [comment]
> Can we make the import_map.json path non-configurable? It's a bit odd that it can be set for serve but not deploy

Yup, I had the same thought after making the PR :) let's do that. So a `--import-map` flag on serve would look for the file on default location.

**Comment by laktek** [comment]
@soedirgo I've updated the PR to make the `import_map.json` path to be non-configurable in serve command.

**Comment by soedirgo** [comment]
Added a commit to:
- remove the option altogether and just omit the option if there's no `import_map.json`
- fix import map path to always use Unix path separator, since the container image is always assumed to be Linux

File: internal/functions/serve/serve.go [hunk 1]
```diff
@@ -94,11 +94,26 @@ func Run(ctx context.Context, slug string, envFilePath string, verifyJWT bool, f
...
 	localFuncDir := filepath.Join(utils.FunctionsDir, slug)
+	localImportMapPath := filepath.Join(localFuncDir, "import_map.json")
+	// We assume the image is always Linux, so path separator must always be `/`.
+	// We can't use filepath.Join because it uses the path separator for the host system, which is `\` for Windows.
 	dockerFuncPath := relayFuncDir + "/" + slug + "/index.ts"
+	dockerImportMapPath := relayFuncDir + "/" + slug + "/import_map.json"
+
+	denoCacheCmd := []string{"deno", "cache"}
+	{
+		if _, err := os.Stat(localImportMapPath); err == nil {
```
Comment (sweatybridge) [inline internal/functions/serve/serve.go:105]:
```suggestion
		if _, err := fsys.Stat(localImportMapPath); err == nil {
```

File: internal/functions/serve/serve.go [hunk 2]
```diff
@@ -125,14 +140,24 @@ func Run(ctx context.Context, slug string, envFilePath string, verifyJWT bool, f
 			}
 		}
 
+		denoRunCmd := []string{"deno", "run", "--no-check=remote", "--allow-all", "--watch", "--no-clear-screen", "--no-npm"}
+		{
+			if _, err := os.Stat(localImportMapPath); err == nil {
```
Comment (sweatybridge) [inline internal/functions/serve/serve.go:145]:
```suggestion
			if _, err := fsys.Stat(localImportMapPath); err == nil {
```

**Review by sweatybridge — CHANGES_REQUESTED** [review]

File: internal/utils/eszip/build.ts [hunk 3]
```diff
@@ -8,27 +8,42 @@ import { load } from "https://deno.land/x/eszip@v0.30.0/loader.ts";
...
-        actualPath = specifier.replace(virtualBasePath, `./${path.dirname(p)}/`);
+        actualPath = specifier.replace(virtualBasePath, `./${funcDirPath}/`);
       }
-      const content = await Deno.readTextFile((actualPath));
-      return {
-        kind: "module",
-        specifier,
-        content
+      try {
+        const content = await Deno.readTextFile((actualPath));
```
Comment (sweatybridge) [inline internal/utils/eszip/build.ts:25]: Is the inner bracket redundant?
```suggestion
        const content = await Deno.readTextFile(actualPath);
```

File: internal/functions/serve/serve.go (hunk 1 above)
Comment (soedirgo) [reply internal/functions/serve/serve.go:105]: Oops, thanks!

**Review by sweatybridge — APPROVED** [review]

**Comment by laktek** [comment]
@sweatybridge @soedirgo Just merge this PR yet. Noticed an issue with Deno Deploy config (which would affect older functions), we probably need to do a small change before releasing this.

**Comment by laktek** [comment]
@soedirgo @sweatybridge I had to add another commit to the PR to send `import_map` param to the API when creating or updating a function. https://github.com/supabase/cli/pull/662/commits/f8279b71cc7ae0961bfbbb483a09fc44f99d4ea5 

This ensures we don't break old functions without an import map.

**Pull request merged**